src/superstruct.ts

```diff
@@ -0,0 +1,44 @@
+import type { Struct } from '@metamask/superstruct';
+import { define } from '@metamask/superstruct';
+
+/**
+ * Defines a new string-struct matching a regular expression.
+ *
+ * Example:
+ *
+ * ```ts
+ * const EthAddressStruct = definePattern('EthAddress', /^0x[0-9a-f]{40}$/iu);
+ * type EthAddress = Infer<typeof EthAddressStruct>; // string
+ *
+ * const CaipChainIdStruct = defineTypedPattern<`${string}:${string}`>(
+ *   'CaipChainId',
+ *   /^[-a-z0-9]{3,8}:[-_a-zA-Z0-9]{1,32}$/u;
+ * );
+ * type CaipChainId = Infer<typeof CaipChainIdStruct>; // `${string}:${string}`
+ *
+ * ```
+ *
+ * @param name - Type name.
+ * @param pattern - Regular expression to match.
+ * @template Pattern - The pattern type, defaults to `string`.
+ * @returns A new string-struct that matches the given pattern.
+ */
+export function definePattern<Pattern extends string = string>(
+  name: string,
+  pattern: RegExp,
+): Struct<Pattern, null> {
+  return define<Pattern>(name, (value: unknown): boolean | string => {
+    const valid = typeof value === 'string' && pattern.test(value);
+
+    if (!valid) {
+      return `Expected a value of type \`${name}\` (pattern \`${
+        pattern.source
+      }\`), but received: \`${
+        // We need to disable this since `value` is `unknown`.
+        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+        value
+      }\``;
+    }
```

Comment:
This will be quite verbose for some CAIP structs, but I don't see this as a problem, e.g:
```ts
Expected a value of type `CaipAssetId` (pattern `^(?<chainId>(?<namespace>[-a-z0-9]{3,8}):(?<reference>[-_a-zA-Z0-9]{1,32}))\\/(?<assetNamespace>[-a-z0-9]{3,8}):(?<assetReference>[-.%a-zA-Z0-9]{1,128})\\/(?<tokenId>[-.%a-zA-Z0-9]{1,78})$`), but received: `foobar`"
```

WDYT?